### ISSUE TITLE
JWlink does not build under MS-DOS 7 (under Win98SE)

### ISSUE THREAD

**Issue by awik32** [Baron-von-Riedesel/jwlink#1]
Hi,

First, the Makefile refers to $(watcompath)/binnt regardless of the host system to be used for building.  Since I installed OpenWatcom for DOS, the binnt directory does not have the necessary files (most notably wcc386.exe).

Second, after changing "binnt" to "binw", and then running wmake, there is an error about "Arg list too big" and the make process is terminated.  Maybe Watcom still (version 1.9) does not support the CMDLINE environment variable for passing command lines of up to 1024 characters?

Regards,
Albert Wik.

**Comment by Baron-von-Riedesel** [comment]
I'm pretty sure I created the jwlink binary in DOS, using the HX DOS extender and an installed Watcom for WinNT. This works, probably because HX does use the CMDLINE environment variable.

**Comment by awik32** [comment]
On Mon, 7 Sep 2020 at 16:25, Andreas Grech <notifications@github.com> wrote:

> I'm pretty sure I created the jwlink binary in DOS, using the HX DOS
> extender and an installed Watcom for WinNT. This works, probably because HX
> does use the CMDLINE environment variable.
>
Can I replace the DOS4GW stub of the binw executables with a HX one?

-aw

**Comment by Baron-von-Riedesel** [comment]
Yes. Just have a look into file "Makefile".

**Comment by awik32** [comment]
On Fri, 11 Sep 2020 at 03:31, Andreas Grech <notifications@github.com> wrote:
>
> Yes. Just have a look into file "Makefile".

C:\hx\tmp>..\bin\pestub -r -v wcc386.exe c:\hx\bin\loadpe.bin
stub file 'c:\hx\bin\loadpe.bin' found
wcc386.exe has no valid PE format - magic bytes PE not found

wcc386.exe appears to be an LE, not PE, executable file.

To work around the problems, I installed the Win32 version of OpenWatcom.

I then found that the build process requires "tinyio.h" which is only
found in the latest release of the OpenWatcom source code, which I
incidentally have installed already.  So I copied the "tinyio.h"...

Then I discovered that the build also requires JWlib.  I found a .zip
file in bin containing a binary-only release of JWlink and JWlib, so I
copied JWlib.exe to watcom/binnt.  Then I had to do the same with
JWlink.exe.

Then I had to copy c:\hx\owsupp\nt\cstrwhx.obj to watcom\lib386\nt.

Then I had to add c:\hx\bin to the PATH so that Wmake would find
pestub.exe.  Finally, the build process completed.

Did you receive the message I sent you on SourceForge about linking
VxDs?  I'm going to test if the GitHub version of JWlink is any
different from the older SourceForge (or your web site) version.

-aw

**Comment by Baron-von-Riedesel** [comment]
> Did you receive the message I sent you on SourceForge about linking VxDs?

Yes. I don't know yet why INT 1's are generated by this version.
I have no working Win9x environment anymore, so to examine why a VxD's real-mode initialization routine isn't called when linked with jwlink might take some time. However, since the cause most likely is just an (uninitialized?) field in the LE header, you can probably find out what's missing.

>  I'm going to test if the GitHub version of JWlink is any different from the older SourceForge (or your web site) version.

Yes, there is a small change, I added the NODOSSEG option for PE binaries. It's not documented yet.

**Comment by awik32** [comment]
> I have no working Win9x environment anymore, ...

If you have some disk space (and a reasonably modern CPU), that is easy to fix.  I'm on Linux 64-bit, and using VMware (there's a free one called "Player") for my Win9x work.  It's quite fast.

> so to examine why a VxD's real-mode initialization routine isn't called when linked with jwlink might take some time. However, since the cause most likely is just an (uninitialized?) field in the LE header, you can probably find out what's missing.

As a matter of fact, the real-mode initialisation issue has been resolved.  With the beta15 version, it works fine (*after* you post-process with the addhdr.exe utility).

The difference between the output of Wlink and JWlink are as follows (both processed with addhdr.exe):
```
Q:\c400\c\20200826\masm>comp bak\wlink\addhdr\masmvxd.386 bak\jwlink\addhdr\masmvxd.386
Comparing BAK\WLINK\ADDHDR\MASMVXD.386 and BAK\JWLINK\ADDHDR\MASMVXD.386...
Compare error at OFFSET 164
file1 = 55
file2 = 15
Compare error at OFFSET 17C
file1 = 45
file2 = 5

Compare more files (Y/N) ?
```
The differences between the raw output of JWlink and the same file after post-processing with addhdr are as follows:
```
Q:\c400\c\20200826\masm>comp bak\jwlink\masmvxd.386 bak\jwlink\addhdr\masmvxd.386
Comparing BAK\JWLINK\MASMVXD.386 and BAK\JWLINK\ADDHDR\MASMVXD.386...
Compare error at OFFSET 140
file1 = 0
file2 = 4D
Compare error at OFFSET 141
file1 = 0
file2 = 41
Compare error at OFFSET 142
file1 = 0
file2 = A
Compare error at OFFSET 143
file1 = 0
file2 = 3

Compare more files (Y/N) ?
```
0x4D is 'M' and 0x41 is 'A', which is the ID of the VxD as used in the DDB struct.  0x30A is the version (3.10) of the DDK (driver development kit) from which the addhdr.exe utility is taken.  So it is very simple.

Do you have documentation on the LE format?

-aw
